lib/fetch-server-test.js

```diff
@@ -0,0 +1,155 @@
+import intersection from 'lodash/intersection';
+import difference from 'lodash/difference';
+
+import * as sinon from 'sinon';
+
+const VALID_METHODS = [`GET`, `POST`, `PATCH`, `PUT`, `DELETE`];
+
+/**
+ * Utility class to stub fetch requests with sinon. Instantiate with a set of
+ * route handlers mapping URL (sub)strings to canned response values, or functions
+ * which return canned responses/response values. Values will be serialized as JSON
+ * and returned in a Fetch-like Promise object.
+ *
+ * @example
+ * const fetchServer = new FetchServer({
+ *   '/foo.json': 'bar',
+ *   '/baz/quux': (url, params) => !!params.foo ? 'a' : 'z',
+ * }, {debug: true});
+ * fetchServer.start();
+ * const res = await fetch('https://myserver/foo.json');
+ * const val = await res.json(); // 'bar'
+ * 
+ * @example 
+ * You can also have the following handler types where key is any of the HTTP methods
+ * const handler = {
+ *   '/foo/bar`: {
+ *      GET: {
+ *        'content-type': 'application/json',
+ *        'status': 200,
+ *        'result': {}
+ *      },
+ *      // Takes a non-async function, and creates Promise
+ *      POST: (url, params) => ({
+ *        'content-type': 'text/html',
+ *        'status': 200,
+ *        'result': {}
+ *      })
+ *   }
+ * }
+ */
+export class FetchServer {
+  constructor(handlers, { debug = false } = {}) {
+    this.handlers = handlers;
+    this.debug = debug;
+    this.handle = this.handle.bind(this);
+  }
+
+  handle(url, params) {
+    // default method GET
+    const HTTPmethod = params.method || `GET`;
+    let response;
+    for (const [route, handler] of Object.entries(this.handlers)) {
+
+      // Block 1 : parses body
+      if (url.includes(route)) {
+        if (params && params.body) {
+          try {
+            params = JSON.parse(params.body);
+          } catch (e) {
+            if (e instanceof SyntaxError) {
+              // not JSON parse-able, must be a string
+              params = params.body;
+            } else {
+              throw new Error(`Invalid body: ${params.body}`);
+            }
+          }
+        }
+
+        // handler is either normal or async function. 
+        if (isFunction(handler)) response = handler(url, params);
+
+        // handler is an object that is either
+        /**
+         * const obj = {
+         *  token: 'xyz',
+         *  token-x: 'xxyz'
+         * }
+         * OR
+         * const obj = {
+         *  GET: {},
+         *  POST: function, // can be async or non-async
+         * }
+         */
+        else {
+          const keys = Object.keys(handler);
+
+          // if keys has all valid http methods          
+          if (intersection(keys, VALID_METHODS).length === keys.length) {
+            for (const [method, methodHandler] of Object.entries(handler)) {
+              if (HTTPmethod === method && method === `GET`) {
+                response = methodHandler;
+              } else if (HTTPmethod === method) {
```

Comment:
I don't think we should restrict whether you pass object or function to these HTTP methods - looks like currently you must pass `object` to GET and `function` to the rest? Could see use cases for GET function handler (i.e. want to generate a response based on parameters in url, id, etc.) or POST with object handler (just want to mock simple `{success: true}` response or something). The API will be simpler to work with if the type of handler and request method are independent.